### PR TITLE
fix(datasource): optimize datasource synchronization

### PR DESCRIPTION
**Pull request by MarkPotato777**
#### What type of PR is this?
type-bug

#### What this PR does / why we need it:
This PR fixed these issues:
1. The individual space databases cannot be displayed in real-time.
2. Oceanbase MySQL datasource synchronization heavily depends on oceanbase.gv$database select privilege.
3. The datasource synchronization is not triggered when modifying the data source.
4. The datasource synchronization schedule is unexpectedly slow.

#### How this PR fixed the above issues respectively:
1. When listing databases, we do a datasource synchronization automatically
2. Use information_schema.schemata instead of oceanbase.gv$database. Additionally, we do not rely on a real unique database_id anymore, we just simply use the database name as the database_id instead.
3. After modifying datasource configuration, we trigger a datasource synchronization.
4. Add the thread pool size to 64.

#### Other changes:
1. Change the timeout of internalProxyRestTemplate from 5s to 10s to avoid possible timeout.

### CODE REVIEW

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_1_6__alter_connect_database.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+alter table `connect_database` modify `charset_name` varchar(128) default null comment 'database charset name';
```
Comment (MarkPotato777) [inline server/odc-migrate/src/main/resources/migrate/common/V_4_2_1_6__alter_connect_database.sql:1]: Actually in individual space, we do not need to obtain charset_name, collation_name, and other information, so I remove the NotNull constraint of these columns.

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_1_6__alter_connect_database.sql [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+alter table `connect_database` modify `collation_name` varchar(128) default null comment 'database collation name';
+alter table `connect_database` modify `table_count` bigint(20) default null comment 'table count which belongs to the database';
+alter table `connect_database` modify `last_sync_time` datetime not null default CURRENT_TIMESTAMP comment 'table count which belongs to the database';
```
Comment (yizhouxw) [inline server/odc-migrate/src/main/resources/migrate/common/V_4_2_1_6__alter_connect_database.sql:4]: comment of field `last_sync_time` incorrect

File: server/odc-service/src/main/java/com/oceanbase/odc/config/RestTemplateConfig.java [hunk 3]
```diff
@@ -63,8 +63,8 @@ public RestTemplate vpcRestTemplate() {
                         MediaType.APPLICATION_FORM_URLENCODED));
         RestTemplate restTemplate = new RestTemplateBuilder()
                 .additionalMessageConverters(stringHttpMc, fastJsonHttpMc)
-                .setConnectTimeout(Duration.ofSeconds(5))
-                .setReadTimeout(Duration.ofSeconds(5))
+                .setConnectTimeout(Duration.ofSeconds(10))
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/config/RestTemplateConfig.java:66]: 5 seconds is suitable for ConnectTimeout

File: server/odc-service/src/main/java/com/oceanbase/odc/config/ScheduleConfiguration.java [hunk 4]
```diff
@@ -168,7 +168,7 @@ public ThreadPoolTaskExecutor cloudLoadDataTaskExecutor() {
     @Bean(name = "syncDatabaseTaskExecutor")
     public ThreadPoolTaskExecutor syncDatabaseTaskExecutor() {
         ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
-        int poolSize = Math.max(SystemUtils.availableProcessors(), 5);
+        int poolSize = 64;
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/config/ScheduleConfiguration.java:171]: avoid hardcode, extract to system configuration, 
or  something like `max (SystemUtils.availableProcessors() * 8,  64)`

**Review by yizhouxw — COMMENTED** [review]
currently the databases sync executing every listDatabases API calling,
how about add min interval ?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/database/DatabaseService.java [hunk 5]
```diff
@@ -371,87 +384,158 @@ public Boolean internalSyncDataSourceSchemas(@NonNull Long dataSourceId) throws
...
+                            .collect(Collectors.groupingBy(DatabaseEntity::getName));
+            List<DatabaseEntity> existedDatabasesInDb =
+                    databaseRepository.findByConnectionId(connection.getId()).stream()
+                            .filter(database -> database.getExisted()).collect(
+                                    Collectors.toList());
+            Map<String, List<DatabaseEntity>> existedDatabaseName2Database =
+                    existedDatabasesInDb.stream().collect(Collectors.groupingBy(DatabaseEntity::getName));
+
+            List<Object[]> toAdd = latestDatabases.stream()
+                    .filter(database -> !existedDatabaseName2Database.keySet().contains(database.getName()))
```
Comment (MarkPotato777) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/database/DatabaseService.java:440]: move existedDatabaseName2Database.keySet() out of iterator

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/database/DatabaseService.java [hunk 6]
```diff
@@ -371,87 +384,158 @@ public Boolean internalSyncDataSourceSchemas(@NonNull Long dataSourceId) throws
...
                     .collect(Collectors.toList());
-            databaseRepository.saveAll(toAdd);
 
-            List<Long> toDelete =
-                    databasesInDb.stream()
-                            .filter(database -> !latestDatabaseId2Database.keySet().contains(database.getDatabaseId()))
-                            .map(DatabaseEntity::getId)
+            JdbcTemplate jdbcTemplate = new JdbcTemplate(dataSource);
+            if (org.apache.commons.collections4.CollectionUtils.isNotEmpty(toAdd)) {
+                jdbcTemplate.batchUpdate(
```
Comment (MarkPotato777) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/database/DatabaseService.java:458]: add transaction?

**Review by PeachThinking — COMMENTED** [review]
Should constraint of `uk_connect_database_connection_id_db_id` on table `connect_database` be deleted?

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_1_6__alter_connect_database.sql (hunk 2 above)
Comment (MarkPotato777) [reply server/odc-migrate/src/main/resources/migrate/common/V_4_2_1_6__alter_connect_database.sql]: fixed

File: server/odc-service/src/main/java/com/oceanbase/odc/config/ScheduleConfiguration.java (hunk 4 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/config/ScheduleConfiguration.java]: fixed

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/database/DatabaseService.java (hunk 5 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/connection/database/DatabaseService.java]: done

**Comment by MarkPotato777** [comment]
> Should constraint of `uk_connect_database_connection_id_db_id` on table `connect_database` be deleted?

Not necessarily, according to the current implementation, database_id would be assigned as a UUID which could be considered as unique.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/DatabaseSyncSchedules.java [hunk 7]
```diff
@@ -44,17 +44,17 @@ public class DatabaseSyncSchedules {
...
             return;
         }
-        for (ConnectionConfig connection : allConnections) {
+        for (ConnectionConfig dataSource : orgDataSources) {
             try {
-                SecurityContextUtils.setCurrentUser(connection.getCreatorId(), connection.getOrganizationId(), null);
-                databaseSyncManager.syncDataSource(connection.getId());
-                log.debug("sync datasource successfully, connectionId={}", connection.getId());
+                databaseSyncManager.submitSyncDataSourceTask(dataSource);
+                log.debug("sync datasource successfully, connectionId={}", dataSource.getId());
```
[the remark this answers is not in the document]
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/connection/DatabaseSyncSchedules.java]: done

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/database/DatabaseService.java (hunk 6 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/connection/database/DatabaseService.java:460]: No need for now as the performance has not yet reached the bottleneck.

File: server/odc-service/src/main/java/com/oceanbase/odc/config/RestTemplateConfig.java (hunk 3 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/config/RestTemplateConfig.java]: fixed

**Review by yizhouxw — APPROVED** [review]
lgtm

**Pull request merged**